app/src/main/java/com/example/weather/MainActivity.java

```diff
@@ -1,14 +1,202 @@
+                        return;
+                    }
+                } catch (JSONException e) {
+                    updateTemperatureView(getText(R.string.error_wrong_city).toString());
+                    Log.d(TAG, "parseTemperature Error");
+                    return;
+                }
+
+                String temperatureTemplate = new String((String) getText(R.string.template_temperature_message));
+                String temperatureMessage = String.format(temperatureTemplate, temperature);
+                updateTemperatureView(temperatureMessage);
+
+                Log.d(TAG, "run finish");
+            }
+        }).start();
+
+        Log.d(TAG, "OnClick finish");
+    }
+
+    public void onClickByLocation(View view) {
+        // todo: implement
+        // https://trello.com/c/W4VxNHog
+    }
+
+    private void updateTemperatureView(String massage) {
+        viewShowWeather.post(new Runnable() {
+            @Override
+            public void run() {
+                viewShowWeather.setText(massage);
+            }
+        });
+    }
+
+    private URL buildRequestUrlWithCity(String cityName) throws MalformedURLException {
+        Uri builtUri = Uri.parse(getText(R.string.weather_api_entry_point).toString())
+                .buildUpon()
+                .appendQueryParameter("q", cityName)
+                .appendQueryParameter("appid", getText(R.string.weather_api_key).toString())
+                .appendQueryParameter("units", "metric")
+                .build();
+        return new URL(builtUri.toString());
+    }
+
+    @Nullable
+    private String parseTemperature(String response) throws JSONException {
+        String temperature = null;
+
+        JSONObject json = new JSONObject(response);
+        JSONObject main  = json.getJSONObject("main");
+        temperature = main.getString("temp");
+
+        return temperature;
+    }
+
+    @Nullable
+    private String doRequest(URL weatherEndpoint) throws IOException {
+        Log.d(TAG, "doRequest start");
+        String respond = null;
+
+        BufferedReader reader = null;
+        InputStream stream = null;
+        HttpsURLConnection connection = null;
+
+        try {
+            connection = (HttpsURLConnection) weatherEndpoint.openConnection();
+            connection.setRequestProperty("User-Agent", getText(R.string.app_name).toString());
+            connection.setRequestMethod("GET");
+            connection.setReadTimeout(3000);
+            connection.connect();
+
+            if (connection.getResponseCode() == HttpsURLConnection.HTTP_OK) {
+                stream = connection.getInputStream();
+                reader = new BufferedReader(new InputStreamReader(stream));
+                StringBuilder buf = new StringBuilder();
+                String line;
+                while ((line = reader.readLine()) != null) {
+                    buf.append(line).append("\n");
+                }
+                respond = buf.toString();
```

Comment:
todo: move out into StreamUtils.streamToString()
```suggestion
                respond = StreamUtils.streamToString()
```